fortran/spglib_f08.F90

```diff
@@ -383,49 +476,229 @@ end function spg_get_error_code
+            type(SpglibMagneticSpacegroupType_c) :: magspgtype_c
+        end function spg_get_magnetic_spacegroup_type_c
+
+        function spg_get_magnetic_dataset_c &
+                & (lattice_c, position_c, types_c, tensors_c, tensor_rank_c, num_atom_c, is_axial_c, symprec_c) &
+                & bind(c, name='spg_get_magnetic_dataset') result(mag_dset_c)
+            import c_int, c_double, c_ptr
+            real(c_double), intent(in) :: lattice_c(3, 3)
+            real(c_double), intent(in) :: position_c(3, *)
+            integer(c_int), intent(in) :: types_c(*)
+            real(c_double), intent(in) :: tensors_c(*)
+            integer(c_int), intent(in), value :: tensor_rank_c
+            integer(c_int), intent(in), value :: num_atom_c
+            integer(c_int), intent(in), value :: is_axial_c
+            real(c_double), intent(in), value :: symprec_c
+            type(c_ptr) :: mag_dset_c
+        end function spg_get_magnetic_dataset_c
+
+        subroutine spg_free_magnetic_dataset_c(dataset) bind(c, name='spg_free_magnetic_dataset')
+            import SpglibMagneticDataset_c
+            type(SpglibMagneticDataset_c), intent(inout) :: dataset
+        end subroutine spg_free_magnetic_dataset_c
+    end interface
+contains
+    ! Copy a C string, passed by pointer, to a Fortran string.
+    ! If the C pointer is NULL, the Fortran string is blanked.
+    ! C_string must be NUL terminated, or at least as long as F_string.
+    ! If C_string is longer, it is truncated. Otherwise, F_string is
+    ! blank-padded at the end.
+    subroutine C_F_string_ptr(C_string, F_string)
+        type(C_ptr), intent(in) :: C_string
+        character(len=*), intent(out) :: F_string
+
+        character(len=1, kind=C_char), dimension(:), pointer :: p_chars
+        integer :: i
+        if (.not. C_associated(C_string)) then
+            F_string = ' '
+        else
+            call C_F_pointer(C_string, p_chars, [huge(0)])
+            i = 1
+            do while (p_chars(i) /= C_NULL_CHAR .and. i <= len(F_string))
+                F_string(i:i) = p_chars(i)
+                i = i + 1
+            end do
+            if (i < len(F_string)) F_string(i:) = ' '
+        end if
+    end subroutine C_F_string_ptr
 
+    ! Copy a C string, passed as a char-array reference, to a Fortran string.
+    subroutine C_F_string_chars(C_string, F_string)
+        character(len=1, kind=C_char), intent(in) :: C_string(*)
+        character(len=*), intent(out) :: F_string
         integer :: i
+        i = 1
+        do while (C_string(i) /= C_NULL_CHAR .and. i <= len(F_string))
+            F_string(i:i) = C_string(i)
+            i = i + 1
+        end do
+        if (i < len(F_string)) F_string(i:) = ' '
+    end subroutine C_F_string_chars
+
+    module function spg_get_version() result(version)
+        character(len=:), allocatable :: version
+
+        call C_F_string(spg_get_version_c(), version)
+    end function spg_get_version
```

Comment:
I am having some trouble with designing this interface. I could probably re-use the implementation in `spg_get_error_message`, but I wanted to design something more modern without fix-sized character array :worried:. @atztogo Got any ideas for modernizing this?

Reply:
Sorry, I have no idea. Is there any problem in what you show above the allocatable string?

Reply:
The issue when I ran the tests is that it detects it does not compare the strings properly. Inside the `C_F_string`, everything is fine and assigned properly, but when it does the comparison afterwards, it reports it as empty. I'm not sure if the issue is in the comparison operation, i.e. `(len=*)`, vs `(len=:)`.

For now I have fixed the size of the strings, and left it to be investigated more afterwards.